spec/core/ics-033-multi-hop/README.md

```diff
@@ -0,0 +1,559 @@
+func VerifyMultihopNonMembership(
+    consensusState exported.ConsensusState,
+    connectionHops []string,
+    proof []byte,
+    prefix exported.Prefix,
+    key string,
+) {
+    // deserialize proof bytes into multihop proofs
+    proofs := abortTransactionUnless(Unmarshal(proof))
+    abortTransactionUnless(len(proofs.ConsensusProofs) >= 1)
+    abortTransactionUnless(len(proofs.ConnectionProofs) == len(proofs.ConsensusProofs))
+
+    // verify connection hop ordering and connections are in OPEN state
+    abortTransactionUnless(VerifyConnectionHops(proofs.ConnectionProofs, connectionHops))
+
+    // verify intermediate consensus and connection states from receiver --> sender
+    abortTransactionUnless(VerifyConsensusAndConnectionStates(consensusState, proofs.ConsensusProofs, proofs.ConnectionProofs))
+
+    // verify a key/value proof on source chain's consensus state.
+    abortTransactionUnless(VerifyKeyNonMembership(proofs, prefix, key))
+}
+
+// VerifyConnectionHops checks that each connection in the multihop proof is OPEN and matches the connections in connectionHops.
+func VerifyConnectionHops(
+    connectionProofs []*ProofData,
+    connectionHops []string,
+) {
+    // check all connections are in OPEN state and that the connection IDs match and are in the right order
+    for i, connData := range connectionProofs {
+        connectionEnd := abortTransactionUnless(Unmarshal(connData.Value))
+
+        // Verify the rest of the connectionHops (first hop already verified)
+        // 1. check the connectionHop values match the proofs and are in the same order.
+        connectionID := parseConnectionID(connData.Key)
+        abortTransactionUnless(connectionID == connectionHops[i+1])
+
+        // 2. check that the connectionEnd's are in the OPEN state.
+        abortTransactionUnless(connectionEnd.GetState() == int32(connectiontypes.OPEN))
+    }
+}
+
+// VerifyConsensusAndConnectionStates verifies the state of each intermediate consensus and connection state
+// starting from the receiving chain and finally proving the sending chain consensus and connection state.
+func VerifyConsensusAndConnectionStates(
+    consensusState exported.ConsensusState,
+    consensusProofs []*ProofData,
+    connectionProofs []*ProofData,
+) {
+    // iterate through proofs to prove from executing chain (receiver) to counterparty chain (sender)
+    var clientID string
+    for i := 0; i < len(consensusProofs); i++ {
+
+        // check the clientID from the previous connection is part of the next
+        // consensus state's key path.
+        abortTransactionUnless(VerifyClientID(clientID, consensusProofs[i].Key))
+
+        // prove the consensus state of chain[i] on chain[i-1]
+        consensusProof := abortTransactionUnless(Unmarshal(consensusProofs[i].Proof))
+        abortTransactionUnless(consensusProof.VerifyMembership(
+            commitmenttypes.GetSDKSpecs(),
```

Comment:
One question:
Does this assume that all proof specs for membership verification for each chain are equal? If not, then I think we need to refer to the client state as well as the consensus state for each chain.

Reply:
Yeah currently it is assuming proof specs are equal.

Reply:
imo this is a very important point and we should make it very clear in the spec.